File: packages/dvai-bridge-capacitor-llama/android/src/test/java/co/deepvoiceai/dvaibridge/llama/HttpServerTest.kt
package co.deepvoiceai.dvaibridge.llama

import kotlinx.coroutines.runBlocking
import org.junit.After
import org.junit.Assert.*
import org.junit.Test

class HttpServerTest {
    private val servers = mutableListOf<HttpServer>()

    @After
    fun tearDown() = runBlocking {
        servers.forEach { runCatching { it.stop() } }
        servers.clear()
    }

    @Test
    fun `bind base port when free`() = runBlocking {
        val server = HttpServer().also { servers.add(it) }
        val port = server.tryBind(basePort = 39101, maxAttempts = 4, host = "127.0.0.1")
        assertEquals(39101, port)
    }

    @Test
    fun `falls back to next port on conflict`() = runBlocking {
        val blocker = HttpServer().also { servers.add(it) }
        blocker.tryBind(basePort = 39110, maxAttempts = 1, host = "127.0.0.1")

        val server = HttpServer().also { servers.add(it) }
        val port = server.tryBind(basePort = 39110, maxAttempts = 4, host = "127.0.0.1")
        assertEquals(39111, port)
    }

    @Test
    fun `stop is idempotent`() = runBlocking {
        val server = HttpServer().also { servers.add(it) }
        // Stop before start — should not throw
        server.stop()
        server.tryBind(basePort = 39120, maxAttempts = 1, host = "127.0.0.1")
        server.stop()
        // Double-stop — should not throw
        server.stop()
    }

    @Test
    fun `throws actionable error when all ports blocked`() = runBlocking {
        @Suppress("UNUSED_VARIABLE")
        val blockers = (0 until 3).map { i ->
            HttpServer().also {
                servers.add(it)
                it.tryBind(basePort = 39130 + i, maxAttempts = 1, host = "127.0.0.1")
            }
        }
        val server = HttpServer().also { servers.add(it) }
        try {
            server.tryBind(basePort = 39130, maxAttempts = 3, host = "127.0.0.1")
            fail("Expected exception")
        } catch (e: IllegalStateException) {
            assertTrue(
                "Error should name the tried range: ${e.message}",
                e.message?.contains("39130..39132") == true,
            )
        }
    }
}

File: packages/dvai-bridge-ios-llama-core/ios/Tests/DVAILlamaCoreTests/HttpServerTest.swift
import XCTest
@testable import DVAILlamaCore

final class HttpServerTest: XCTestCase {
    func testTryBindBindsBasePort() async throws {
        let server = HttpServer()
        let port = try await server.tryBind(basePort: 39001, maxAttempts: 4, host: "127.0.0.1")
        XCTAssertEqual(port, 39001)
        await server.stop()
    }

    func testTryBindFallsBackOnPortInUse() async throws {
        // Block port 39010 with another server
        let blocker = HttpServer()
        _ = try await blocker.tryBind(basePort: 39010, maxAttempts: 1, host: "127.0.0.1")

        let server = HttpServer()
        let port = try await server.tryBind(basePort: 39010, maxAttempts: 4, host: "127.0.0.1")
        XCTAssertEqual(port, 39011)
        await server.stop()
        await blocker.stop()
    }

    func testStopIsIdempotent() async throws {
        let server = HttpServer()
        await server.stop() // before start — should not throw
        _ = try await server.tryBind(basePort: 39020, maxAttempts: 1, host: "127.0.0.1")
        await server.stop()
        await server.stop() // double-stop should not throw
    }

    func testThrowsActionableErrorWhenAllPortsBlocked() async throws {
        // Block 39030..39032
        var blockers: [HttpServer] = []
        for i in 0..<3 {
            let s = HttpServer()
            _ = try await s.tryBind(basePort: 39030 + i, maxAttempts: 1, host: "127.0.0.1")
            blockers.append(s)
        }
        defer {
            Task { for b in blockers { await b.stop() } }
        }

        let server = HttpServer()
        do {
            _ = try await server.tryBind(basePort: 39030, maxAttempts: 3, host: "127.0.0.1")
            XCTFail("should have thrown")
        } catch let error as NSError {
            XCTAssertTrue(error.localizedDescription.contains("39030..39032"),
                          "Error should name the tried range; got: \(error.localizedDescription)")
        }
    }
}
